optimizes process_cell() and related atmos procs and reduces atmos machine processing

Kylerace opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request
optimizes a few important atmos procs called from process_cell() as well as documents a few things. also makes scrubbers only process on changed gas mixes so they dont waste processing time on gas mixes they cant do anything with. also makes centcom atmos machines start off because theyre a complete waste of cpu. 
<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game
i want to give maptick an extra 5% of the tick at max to lower time dilation, but to do that i want to make subsystems be able to do more with less. the most sensitive subsystem to being choked out is atmos, so im starting with that. atmos now spends less time processing atmos machines since scrubbers only process when gas changes and meters will only change greyscale if they actually change so theyre less intensive. also should scale slightly better with number of active turfs but i still need to benchmark it
<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->


<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

Kylerace committed:
optimizes process_cell() and related atmos procs and adds docs

Kylerace committed:
optimizes /datum/gas_mixture/proc/compare() some more

Kylerace committed:
makes _RemoveFromParent() legible

Kylerace committed:
makes vent scrubbers only activate on atmos changes

Kylerace committed:
makes the vents in centcom start off

Kylerace committed:
makes meters only update appearance when its changed

Kylerace committed:
fixes vent scrubbers and makes them not process on SSmachines

Kylerace committed:
Merge remote-tracking branch 'upstream/master' into process_cell_op

Kylerace commented:
TODO: benchmark it

MrStonedOne commented on code/controllers/subsystem/air.dm:
```diff
@@ -222,8 +222,8 @@ SUBSYSTEM_DEF(air)
 	var/list/currentrun = src.currentrun
-	while(currentrun.len)
-		var/datum/thing = currentrun[currentrun.len]
+	while(length(currentrun))
```
if currentrun goes null this change would hide it where as the old version would runtime.

Things to think about.

Kylerace replied:
it will still runtime on line 227

MrStonedOne replied:
227 wouldn't run because `length(currentrun)` on null returns 0 and this will fail the while check

Only a byond bug would lead to currentrun being null anyways, so this a moot point, but if the copy() fails on a valid list for some reason, the observed behavior is that it would just fail the while check and then reach the end of the proc.

Kylerace replied:
ah yeah youre right

LemonInTheDark commented on code/__DEFINES/atmospherics/atmos_core.dm:
```diff
@@ -101,10 +101,10 @@
-#define MINIMUM_AIR_TO_SUSPEND (MOLES_CELLSTANDARD*MINIMUM_AIR_RATIO_TO_SUSPEND)
-/// Either this must be active (round about 0.1) //Might need to raise this a tad to better support space leaks. we'll see
-#define MINIMUM_MOLES_DELTA_TO_MOVE (MOLES_CELLSTANDARD*MINIMUM_AIR_RATIO_TO_MOVE)
+/// Minimum amount of moles that has to move before a whole-group processing gets delayed (its about 10)
```
? May be better to just say equalizing directly

LemonInTheDark commented on code/__DEFINES/atmospherics/atmos_core.dm:
```diff
@@ -101,10 +101,10 @@
-#define MINIMUM_MOLES_DELTA_TO_MOVE (MOLES_CELLSTANDARD*MINIMUM_AIR_RATIO_TO_MOVE)
+/// Minimum amount of moles that has to move before a whole-group processing gets delayed (its about 10)
+#define MINIMUM_AIR_TO_SUSPEND (MOLES_CELLSTANDARD * MINIMUM_AIR_RATIO_TO_SUSPEND)
+/// Either this must be active (its about 0.1) //Might need to raise this a tad to better support space leaks. we'll see
+#define MINIMUM_MOLES_DELTA_TO_MOVE (MOLES_CELLSTANDARD * MINIMUM_AIR_RATIO_TO_MOVE)
```
Anti (Round about) posters, disgusting

LemonInTheDark commented on code/__DEFINES/atmospherics/atmos_helpers.dm:
```diff
@@ -32,10 +32,10 @@
 ///Directly adds a gas to a gas mixture without checking for its presence beforehand, use only if is certain the absence of said gas
 #define ADD_GAS(gas_id, out_list)\
-	var/list/tmp_gaslist = GLOB.gaslist_cache[gas_id]; out_list[gas_id] = tmp_gaslist.Copy();
+	var/static/list/list/_add_gas_gaslist_cache = GLOB.gaslist_cache; out_list[gas_id] = _add_gas_gaslist_cache[gas_id].Copy();
```
This name is way too verbose
Just call it _static_gaslist_cache, or even just _gaslist_cache

LemonInTheDark commented on code/__DEFINES/atmospherics/atmos_helpers.dm:
```diff
@@ -32,10 +32,10 @@
-///Adds a gas to a gas mixture but checks if is already present, faster than the same proc
-#define ASSERT_GAS(gas_id, gas_mixture) if (!gas_mixture.gases[gas_id]) { ADD_GAS(gas_id, gas_mixture.gases) };
+///Adds a gas to a gases list but checks if is already present, faster than the same proc
+#define ASSERT_GAS(gas_id, gas_mixture_gases) if (!gas_mixture_gases[gas_id]) { ADD_GAS(gas_id, gas_mixture_gases) };
```
Just call it gas_list, that's what it is, and it's fine to refer to it as such

LemonInTheDark commented on code/modules/atmospherics/environmental/LINDA_turf_tile.dm:
```diff
@@ -269,24 +270,28 @@
 			if(our_excited_group != enemy_excited_group)
 				//combine groups (this also handles updating the excited_group var of all involved turfs)
 				our_excited_group.merge_groups(enemy_excited_group)
-				our_excited_group = excited_group //update our cache
```
This isn't workable. We use our_excited_group later on, and we have no guarantee that ours is the excited group that will survive the merge

LemonInTheDark commented on code/modules/atmospherics/environmental/LINDA_turf_tile.dm:
```diff
@@ -269,24 +270,28 @@
 		if(our_excited_group && enemy_excited_group && enemy_tile.excited) //If you're both excited, no need to compare right?
 			should_share_air = TRUE
-		else if(our_air.compare(enemy_air)) //Lets see if you're up for it
-			SSair.add_to_active(enemy_tile) //Add yourself young man
+
+		else if(our_air.compare(enemy_air)) //check if theres enough of a difference to share
+			SSair.add_to_active(enemy_tile)
```
Documentation is for cowards

LemonInTheDark commented on code/modules/atmospherics/environmental/LINDA_turf_tile.dm:
```diff
@@ -323,7 +328,7 @@
 	our_air.react(src)
 
 	update_visuals()
-	if(!consider_superconductivity(starting = TRUE) && !active_hotspot) //Might need to include the return of react() here
+	if(!consider_superconductivity(TRUE) && !active_hotspot) //Might need to include the return of react() here
```
This is a decent hit to readability, a profile of just this would be appreciated

LemonInTheDark commented on code/modules/atmospherics/environmental/LINDA_turf_tile.dm:
```diff
@@ -511,6 +529,18 @@
 		display_id = wrapping_id
 	thing.vis_contents += GLOB.colored_turfs[display_id]
 
+/datum/excited_group/proc/debug_total_moles()
+	var/moles = 0
+	for(var/turf/open/contained_turf as anything in turf_list)
+		var/delta_moles = 0
+		if(!contained_turf.air || !istype(contained_turf.air, /datum/gas_mixture/turf))
+			continue
+
+		TOTAL_MOLES(contained_turf.air.gases, delta_moles)
+		moles += delta_moles
+
+	message_admins("this excited group contained [length(turf_list)] turfs, the combined total of which contained [moles] moles of gas")
+
```
??? (undocced change)

Kylerace replied:
oh right i was fucking around with excited groups and forgot to get rid of this

LemonInTheDark commented on code/modules/atmospherics/gasmixtures/gas_mixture.dm:
```diff
@@ -335,20 +335,14 @@ GLOBAL_LIST_INIT(gaslist_cache, init_gaslist_cache())
-	var/old_sharer_heat_capacity = 0
-	if(abs_temperature_delta > MINIMUM_TEMPERATURE_DELTA_TO_CONSIDER)
-		old_self_heat_capacity = heat_capacity()
-		old_sharer_heat_capacity = sharer.heat_capacity()
+	var/temperature_delta = abs(temperature_archived - sharer.temperature_archived)
 
```
This isn't an optimization

LemonInTheDark commented on code/modules/atmospherics/gasmixtures/gas_mixture.dm:
```diff
@@ -357,32 +351,36 @@ GLOBAL_LIST_INIT(gaslist_cache, init_gaslist_cache())
 	//THERMAL ENERGY TRANSFER
-	if(abs_temperature_delta > MINIMUM_TEMPERATURE_DELTA_TO_CONSIDER)
+	if(temperature_delta > MINIMUM_TEMPERATURE_DELTA_TO_CONSIDER)
+		var/old_self_heat_capacity = heat_capacity(ARCHIVE)
+		var/old_sharer_heat_capacity = sharer.heat_capacity(ARCHIVE)
+
```
It significantly changes behavior. The heat capacity of a share is not the same as the heat capacity of a series of shares sampled at any time

LemonInTheDark commented on code/modules/atmospherics/gasmixtures/gas_mixture.dm:
```diff
@@ -357,32 +351,36 @@ GLOBAL_LIST_INIT(gaslist_cache, init_gaslist_cache())
 
-		var/delta = QUANTIZE(gas[ARCHIVE] - sharergas[ARCHIVE])/(atmos_adjacent_turfs+1) //the amount of gas that gets moved between the mixtures
+		///the amount of gas that gets moved between the mixtures.
+		var/gas_delta = QUANTIZE(gas[ARCHIVE] - sharergas[ARCHIVE]) / (atmos_adjacent_turfs + 1)
```
why rename delta?

Kylerace replied:
was more decriptive imo

LemonInTheDark replied:
I'm gonna be a boomer but I think the code's cleaner with it just being delta.
Maybe I'm dumb. idk

Kylerace replied:
well i think the bikeshed should be red. ill change it back tomorrow

Kylerace replied:
renamed

LemonInTheDark commented on code/modules/atmospherics/machinery/components/unary_devices/vent_scrubber.dm:
```diff
@@ -135,7 +140,24 @@
 	. = ..()
 
-/obj/machinery/atmospherics/components/unary/vent_scrubber/process_atmos()
+/obj/machinery/atmospherics/components/unary/vent_scrubber/should_atmos_process(datum/gas_mixture/air, exposed_temperature)
+	if(welded || !is_operational)
+		return FALSE
+	if(!nodes[1] || !on || !filter_types)
+		on = FALSE
+		return FALSE
+
+	var/list/changed_gas = air?.gases
+
+	if(!changed_gas)
+		return FALSE
+
+	if(scrubbing == SIPHONING || length(filter_types & changed_gas))
+		return TRUE
+
+	return FALSE
+
+/obj/machinery/atmospherics/components/unary/vent_scrubber/atmos_expose(datum/gas_mixture/air, exposed_temperature)
 	if(welded || !is_operational)
 		return FALSE
```
This on its own is not enough. `should_atmos_process` will only be called automatically if the scrubber's turf's mix changes. You will need to manually call it when anything else that might trigger a valid scrub changes

Kylerace replied:
oh yeah i forgot air has lower wait than machines. i still dont like having it in two subsystem processes though when its only ever used here so ill add a cooldown

LemonInTheDark commented on code/modules/atmospherics/machinery/components/unary_devices/vent_scrubber.dm:
```diff
@@ -146,6 +168,7 @@
 		return
 	scrub(us)
 	if(widenet)
+		check_turfs()
```
This seems mildly expensive, though I guess it'll be better now

Kylerace replied:
oh yeah i forgot air has lower wait than machines. i still dont like having it in two subsystem processes though when its only ever used here so ill add a cooldown

Kylerace replied:
done'd

LemonInTheDark reviewed (changes requested):
It's a bit hard to sus out exactly what changes were optimizations, and what was just renames or cleanups of unrelated code. I'd appreciate it if you documented what you changed and why in your pr desc, for the sake of clarity. It also makes double checking your thought processes easier, rather then me needing to just guess what you were going for.

I'd also like to see the specific results of a few of these, rather then lumping them all up into one before/after profile
I realize that's a bit of a tall order, but some of these things are very minor, and some aren't. Having the context of the exact benefits would be useful for them.

Kylerace committed:
creates setters for filter_types to evaluate processing for scrubbers

Kylerace committed:
makes vent_scrubber's check processing on all relevant settings changes

Kylerace committed:
does the rest of lemons reviews i think idr

Kylerace committed:
another lemon review

LemonInTheDark commented:
Ready? Sans merge conflicts

Kylerace committed:
Merge remote-tracking branch 'upstream/master' into process_cell_op

Kylerace committed:
renames gas_delta back to delta and adds a cooldown for checking turfs

Kylerace commented:
@LemonInTheDark ready

Iamgoofball commented:
atmos is your issue, no GBP

LemonInTheDark commented:
The fuck?

LemonInTheDark commented on code/controllers/subsystem/air.dm:
```diff
@@ -535,7 +535,7 @@ SUBSYSTEM_DEF(air)
 		EG.add_turf(src)
 
-	for (var/turf/open/ET in atmos_adjacent_turfs)
+	for (var/turf/open/ET as anything in atmos_adjacent_turfs)
```
This does actually kinda need to be here, adjacencies aren't always rebuilt instantly, because of an attempted explosion optimization, so we can't be guaranteed of the typing of this list

LemonInTheDark commented on code/modules/atmospherics/environmental/LINDA_turf_tile.dm:
```diff
@@ -279,24 +280,29 @@
 		//cache for sanic speed
 		var/datum/excited_group/enemy_excited_group = enemy_tile.excited_group
-		//If we are both in an excited group, and they aren't the same, merge.
+		//If we are both in an excited group, and they aren't the same, merge your group into ours
```
This comment isn't accurate, merging isn't really directional, it goes based off the action that takes the least amount of work

LemonInTheDark commented on code/modules/atmospherics/environmental/LINDA_turf_tile.dm:
```diff
@@ -279,24 +280,29 @@
 				//combine groups (this also handles updating the excited_group var of all involved turfs)
 				our_excited_group.merge_groups(enemy_excited_group)
-				our_excited_group = excited_group //update our cache
+				our_excited_group = excited_group
```
Keep the comment?

LemonInTheDark commented on code/modules/atmospherics/environmental/LINDA_turf_tile.dm:
```diff
@@ -358,15 +364,15 @@
 /turf/open/proc/consider_pressure_difference(turf/target_turf, difference)
+	if(difference < pressure_difference)
+		return
+
 	SSair.high_pressure_delta |= src
```
I think? there was a reason for this, it doesn't come immediately to mind. bit uneasy about this

LemonInTheDark commented on code/modules/atmospherics/gasmixtures/gas_mixture.dm:
```diff
@@ -452,30 +453,24 @@ GLOBAL_LIST_INIT(gaslist_cache, init_gaslist_cache())
-			delta > gas_moles * MINIMUM_AIR_RATIO_TO_MOVE)
+		var/gas_moles = cached_gases[id]?[MOLES]
+
+		var/delta = abs(gas_moles - sample_gases[id]?[MOLES])
```
Will a null behave like 0 here?

Kylerace replied:
yes

LemonInTheDark commented on code/modules/atmospherics/gasmixtures/gas_mixture.dm:
```diff
@@ -492,15 +487,15 @@ GLOBAL_LIST_INIT(gaslist_cache, init_gaslist_cache())
 	for(var/gas_id in cached_gases)
 		var/list/reaction_set = gas_reactions[gas_id]
 		if(!reaction_set)
 			continue
-		pre_formation += reaction_set[1]
-		mid_formation += reaction_set[2]
-		post_formation += reaction_set[3]
-		fires += reaction_set[4]
+		pre_formation += reaction_set[PRIORITY_PRE_FORMATION]
+		mid_formation += reaction_set[PRIORITY_FORMATION]
+		post_formation += reaction_set[PRIORITY_POST_FORMATION]
+		fires += reaction_set[PRIORITY_FIRE]
 
```
Thank you

LemonInTheDark commented on code/modules/atmospherics/machinery/components/unary_devices/vent_scrubber.dm:
```diff
@@ -57,6 +64,65 @@
+
+	var/turf/open/our_turf = get_turf(src)
+	var/datum/gas_mixture/turf_gas = our_turf?.air
+	if(!our_turf || !turf_gas)
```
couldn't this just be turf_gas

LemonInTheDark commented on code/modules/atmospherics/machinery/components/unary_devices/vent_scrubber.dm:
```diff
@@ -57,6 +64,65 @@
+
+	var/turf/open/our_turf = get_turf(src)
+	var/datum/gas_mixture/turf_gas = our_turf?.air
+	if(!our_turf || !turf_gas)
```
ditto

LemonInTheDark commented on code/modules/atmospherics/machinery/components/unary_devices/vent_scrubber.dm:
```diff
@@ -57,6 +64,65 @@
+	if(!our_turf || !turf_gas)
+		return FALSE
+
+	check_atmos_process(src, turf_gas, turf_gas.temperature)
```
source is the turf isn't it?

Kylerace replied:
no

LemonInTheDark replied:
shouldn't it be? that's what the signal sends

Kylerace replied:
i am dumb nm

LemonInTheDark commented on code/modules/atmospherics/machinery/components/unary_devices/vent_scrubber.dm:
```diff
@@ -135,7 +201,24 @@
+		on = FALSE
+		return FALSE
+
+	var/list/changed_gas = air?.gases
```
you can rely on the gases list existing, unless I'm missing something

LemonInTheDark reviewed (changes requested):
Lemme know if I'm missing anything here, but there's an optimization to compare, react, and gas addition, alongside making scrubbers and meters more event based, right?

Kylerace committed:
does some of lemons reviews

Kylerace committed:
Merge remote-tracking branch 'upstream/master' into process_cell_op

Kylerace committed:
makes scrubbers remove small gas amounts better

this is because scrubbers are HORRIBLE at it and now it actually matters from a performance standpoint because scrubbers will go to sleep much sooner

Kylerace committed:
Merge remote-tracking branch 'upstream/master' into process_cell_op

Kylerace commented:
this is what i get for not atomizing my changes ill be back with just the machine stuff

LemonInTheDark commented:
JERRRYYYYYY
